### PR TITLE
1430 as a user my transfer data should not have gaps

### PR DESCRIPTION
**Pull request by preethamr**
Queries form cartographer to subgraphs use timestamps to checkpoint. This checkpoint mechanism because of quirks of how queries to subgraph work, are causing data gaps.

Adding a new mechanism to monitor for data gaps, and fetch from subgraph, to close the gaps.

## Type of change

- [ ] Docs change / dependency upgrade
- [ ] Configuration / tooling changes
- [x] Refactoring
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires changes in customer code

## Related Issue(s)

Fixes 
https://github.com/connext/nxtp/issues/1430
https://github.com/connext/nxtp/issues/1324

### CODE REVIEW

File: packages/adapters/subgraph/src/lib/operations/queries.ts [hunk 1]
```diff
@@ -368,19 +406,52 @@ export const getDestinationTransfersByExecuteTimestampQuery = (
...
+  const domains = Object.keys(config.sources);
+  for (const domain of domains) {
+    const prefix = config.sources[domain].prefix;
+    if (params.has(domain)) {
+      combinedQuery += destinationTransfersByIDsQueryString(
+        prefix,
+        params.get(domain)!.transferIDs.map((id) => `"${id}"`),
+      );
+    } else {
+      console.log(`No agents for domain: ${domain}`);
```
Comment (rhlsthrm) [inline packages/adapters/subgraph/src/lib/operations/queries.ts:449]: Prefer to not use console logs here.

File: packages/adapters/subgraph/src/lib/operations/queries.ts [hunk 2]
```diff
@@ -294,72 +306,98 @@ export const getOriginTransfersQuery = (agents: Map<string, SubgraphQueryMetaPar
...
-  return `
-  ${prefix}_destinationTransfers(
-    where: { 
-      executedTimestamp_gte: ${fromTimestamp}, 
-      ${maxBlockNumber ? `, executedBlockNumber_lte: ${maxBlockNumber}` : ""} 
-    }, 
-    orderBy: executedTimestamp, 
+  return `${prefix}_destinationTransfers(
+    where: {
+      nonce_gte: ${fromNonce},
```
Comment (rhlsthrm) [inline packages/adapters/subgraph/src/lib/operations/queries.ts:358]: The main problem we run into here is that the reconciled transfers don't have a nonce. They only have a timestamp.

File: packages/adapters/subgraph/src/lib/operations/queries.ts [hunk 3]
```diff
@@ -154,9 +159,16 @@ export const getAssetBalancesQuery = (prefix: string, router: string): string =>
...
-export const getAssetBalancesAllRoutersQuery = (prefix: string): string => {
+export const getAssetBalancesRoutersQuery = (
+  prefix: string,
+  offset: number,
+  orderDirection: "asc" | "desc" = "desc",
+): string => {
   const queryString = `
-    ${prefix}_routers {
+    ${prefix}_routers(
+    skip: ${offset},
```
Comment (rhlsthrm) [inline packages/adapters/subgraph/src/lib/operations/queries.ts:169]: We should also use the `first` param here so we can be explicit about how big the pages are.

File: packages/agents/cartographer/poller/db/schema.sql [hunk 4]
```diff
@@ -182,6 +215,13 @@ ALTER TABLE ONLY public.transfers
     ADD CONSTRAINT transfers_pkey PRIMARY KEY (transfer_id);
 
 
+--
+-- Name: transfers update_time_on_transfers; Type: TRIGGER; Schema: public; Owner: -
+--
+
+CREATE TRIGGER update_time_on_transfers BEFORE UPDATE ON public.transfers FOR EACH ROW EXECUTE FUNCTION public.trigger_set_timestamp();
```
Comment (rhlsthrm) [inline packages/agents/cartographer/poller/db/schema.sql:222]: Love it.

File: packages/agents/cartographer/poller/src/lib/operations/transfers.ts [hunk 5]
```diff
@@ -31,58 +49,122 @@ export const updateTransfers = async () => {
...
-    const transfers = await subgraph.getOriginTransfersByXCallTimestamp(subgraphXCallQueryMetaParams);
-    logger.info("Retrieved origin transfers by xcalled timestamp", requestContext, methodContext, {
+    const transfers = await subgraph.getOriginTransfersByNonce(subgraphOriginQueryMetaParams);
+    logger.info("Retrieved origin transfers by nonce", requestContext, methodContext, {
       transfers,
       count: transfers.length,
     });
     await database.saveTransfers(transfers);
+
+    for (const domain of domains) {
```
Comment (rhlsthrm) [inline packages/agents/cartographer/poller/src/lib/operations/transfers.ts:102]: Nit, but this should be able to be parallelized i.e.

```ts
await Promise.all(domains.map(async (domain) => {
  ...
});
```

File: packages/agents/cartographer/poller/src/lib/operations/transfers.ts [hunk 6]
```diff
@@ -31,58 +49,122 @@ export const updateTransfers = async () => {
...
+  if (subgraphDestinationQueryMetaParams.size > 0) {
+    // Get destination transfers for all domains in the mapping.
+    const transfers = await subgraph.getDestinationTransfersByNonce(subgraphDestinationQueryMetaParams);
+    logger.info("Retrieved destination transfers by nonce", requestContext, methodContext, {
       transfers,
       count: transfers.length,
     });
     await database.saveTransfers(transfers);
+
+    for (const domain of domains) {
```
Comment (rhlsthrm) [inline packages/agents/cartographer/poller/src/lib/operations/transfers.ts:121]: Also can be parallelized.

File: packages/adapters/subgraph/src/lib/operations/queries.ts (hunk 2 above)
Comment (preethamr) [reply packages/adapters/subgraph/src/lib/operations/queries.ts:358]: True, we will not get reconciled but not executed transfers here.
They would be fetched by the reconciled timestamp queries, separately

We could use timestamps everywhere, but using nonce is more robust, as it handles the case where there are more transfers that got executed in 1 second, than can fit in response, and cause us repeatedly process then same transfers and get stuck.

So, thought use nonce when we can, if not use timestamp.

Comment (rhlsthrm) [reply packages/adapters/subgraph/src/lib/operations/queries.ts:358]: This makes sense and is likely the best approach for now.

File: packages/agents/cartographer/poller/src/lib/operations/transfers.ts (hunk 5 above)
Comment (preethamr) [reply packages/agents/cartographer/poller/src/lib/operations/transfers.ts]: Yes!

File: packages/agents/cartographer/poller/src/lib/operations/transfers.ts (hunk 6 above)
Comment (preethamr) [reply packages/agents/cartographer/poller/src/lib/operations/transfers.ts]: Done

File: packages/adapters/subgraph/src/lib/operations/queries.ts (hunk 1 above)
Comment (preethamr) [reply packages/adapters/subgraph/src/lib/operations/queries.ts]: Removed them.

File: packages/adapters/subgraph/src/lib/operations/queries.ts (hunk 3 above)
Comment (preethamr) [reply packages/adapters/subgraph/src/lib/operations/queries.ts:170]: Changed it use first as well, with explicit page size.

**Pull request merged**